### ISSUE TITLE
[pytorch] cudnn benchmark=True overrides deterministic=True 

### ISSUE THREAD

**Issue by soumith** [pytorch/pytorch#6351]
currently if in user code, this exists:

```
cudnn.benchmark = True
cudnn.deterministic = True
```

Then, deterministic is silently ignored. Either a Warning or Error should be generated in this case.
reported by @lvdmaaten

**Comment by apaszke** [comment]
Why can’t they coexist? IIRC we use the deterministic flag to filter results of benchmark somehow, isn’t that the case?

**Comment by lvdmaaten** [comment]
I haven't checked the implementation, but what I do know is that currently the outputs of the modules are non-deterministic when `benchmark = True`, even though `deterministic = True`. (Which makes sense as the implementations being used under the hood may vary because of the benchmarking.) That seems undesired: it is promising the user deterministic behavior, but not providing it.

**Comment by ngimel** [comment]
There is logic to filter deterministic algorithms from cudnnFind return results https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cudnn/Conv.cpp#L460-L475 (if benchmark is true). If this is not working, probably cudnn is not returning determinism flag correctly. @lvdmaaten, a small repro would help to isolate if it is cudnn bug or if something is not working correctly in the pytorch bindings.
When benchmark is false, algos are explicitly set to something known to be deterministic (DEFAULT_ALGO), so without benchmark determinism is working.

**Comment by lvdmaaten** [comment]
@ngimel That logic seems wrong to me. Only benchmarking deterministic implementations does not lead to deterministic behavior: two different implementations are very likely to produce different results, even if those implementations themselves are deterministic. (And which implementation is selected by the benchmarking may vary per run, platform, etc.)

Afaict the only way to produce fully deterministic behavior is by not doing any benchmarking when `cudnn.deterministic=True`. If also `cudnn.benchmark=True`, the user should be warned that benchmarking is not possible in a deterministic implementation.

**Comment by ngimel** [comment]
@lvdmaaten You are right, that check is not guaranteeing much, only that a deterministic algorithm would be picked for a run, and another deterministic algo can be picked for another run. However, it is a legitimate question what guarantees should be provided. Even if a particular algo is chosen (like it is currently done for deterministic w/o benchmarking), it does not guarantee same results on the different architectures or between the different cudnn versions or cublas versions.
On the same hardware with the same software stack it should be possible to pick deterministic algos without sacrificing performance in most cases, but that would likely require a user-level API directly specifying algo (lua torch had that), or reimplementing cudnnFind within a framework, like tensorflow does, because the way cudnnFind is implemented it is prone to selecting different algos on the different runs if their runtime is close, because all algos are run just once and variance is high.

**Comment by apaszke** [comment]
That makes sense, thanks for the description @lvdmaaten. Maybe let's raise a warning when both are set for now, and we can figure out more robust ways to benchmark later?

**Comment by lvdmaaten** [comment]
@ngimel Yeah, those are good points. The context for this issue was I was writing some CI tests for my project, checking the output of a small training pipeline. I couldn't get the tests to pass even though `cudnn.deterministic=True` because there was `cudnn.benchmark=True` lurking somewhere in the code. I think a reasonable guarantee would be for `cudnn.deterministic=True` to always produce the same results on the same GPU hardware and the same CUDA / cudnn / cublas software stack -- as this is a minimal requirement for automated tests, and as you say, it seems unlikely we can give any stronger guarantees.

@apaszke Sounds good!

**Comment by dlmacedo** [comment]
I am using:

cudnn.benchmark = False
cudnn.deterministic = True

random.seed(1)
numpy.random.seed(1)
torch.manual_seed(1)
torch.cuda.manual_seed(1)

And still not deterministic behavior…

**Comment by dlmacedo** [comment]
[](https://github.com/dlmacedo/deep-learning-class/blob/master/no_deterministic.py)

**Comment by dlmacedo** [comment]
The following code is NOT deterministic...

I believe maybe is not just a question of setting 

> cudnn.benchmark = False

 and 

> cudnn.deterministic = True

[https://github.com/dlmacedo/deep-learning-class/blob/master/no_deterministic.py](url)

**Comment by apaszke** [comment]
@dlmacedo This is 600 lines of code that uses a lot of other libraries as well. We can take a look at it once you reduce it to a small self contained snippet (let’s say of ~40 lines) that would still exhibit nondeterminism.

**Comment by dlmacedo** [comment]
The problem is in the data augmentation transformations:

```
transforms.RandomCrop(32, padding=4),
transforms.RandomHorizontalFlip(),
```

After I removed the above code, it worked 100% deterministically...

I am using Pytorch 0.3.1

**Comment by dlmacedo** [comment]
This appears to be a similar issue:

https://discuss.pytorch.org/t/non-reproducible-result-with-gpu/1831/5

**Comment by wandering007** [comment]
@dlmacedo I think it is caused by multiple workers initialized with different seeds (https://github.com/pytorch/pytorch/issues/3880).   
You can remove the randomness by setting `args.workers` to 1. Or you can initialize the workers with the same seed, which could probably be deterministic (not 100%).

**Comment by dlmacedo** [comment]
I used `worker_init_fn=worker_init` in the Dataloader and the following function:

```
def worker_init(worker_id):
    torch.manual_seed(worker_id)

```
But it did not work to num_workers>0.

Moreover, I do not think this should be necessary from the point of view of using users simplicity.

After using `cudnn.deterministic = True` and global manual seed, I think it should not be necessary to do other workarounds.

**Comment by dlmacedo** [comment]
@wandering007 It worked. Thanks.

Nevertheless, I think the idea is to have a more straightforward way to get determinist behave. This "workaround" in the works is not so easy to beginners.

**Comment by fmassa** [comment]
@dlmacedo this non-determinism that you are facing has been fixed in pytorch 0.4

**Comment by dlmacedo** [comment]
Based on my tests, even in PyTorch 0.4 we still need to initialize the workers with the same seed to get deterministic behavior. The following lines are NOT enough:

```
cudnn.benchmark = False
cudnn.deterministic = True

random.seed(1)
numpy.random.seed(1)
torch.manual_seed(1)
torch.cuda.manual_seed(1)
```

I think this should not be the standard behavior. In my opinion, the above lines should be enough to provide deterministic behavior. It is not obvious to the novice that, besides the above lines, he also need to initialize the workers with the same seed to get deterministic behavior.

**Comment by SsnL** [comment]
@dlmacedo The pytorch seed is already set properly in workers. It should give you deterministic behavior. If you use np in the dataset getitem, you should set np seed in `init_fn` because I don't think we should assume that np exists.

**Comment by dlmacedo** [comment]
I am NOT using numpy to dataloader.

Since I set the PYTHON seed in my code and it is changed inside dataloader code (see bellow), I need to overwrite the PYTHON seed back again (since it is being changed inside the dataloader code) in the worker init (I am not changing NUMPY seed inside the worker_init):

```
def worker_init(worker_id):
    random.seed(args.base_seed)

```

See below the dataloader code that changes the PYTHON (NOT numpy) seed:

```
def _worker_loop(dataset, index_queue, data_queue, collate_fn, seed, init_fn, worker_id):
    global _use_shared_memory
    _use_shared_memory = True

    # Intialize C side signal handlers for SIGBUS and SIGSEGV. Python signal
    # module's handlers are executed after Python returns from C low-level
    # handlers, likely when the same fatal signal happened again already.
    # https://docs.python.org/3/library/signal.html Sec. 18.8.1.1
    _set_worker_signal_handlers()

    torch.set_num_threads(1)
    random.seed(seed)
    torch.manual_seed(seed)

    if init_fn is not None:
        init_fn(worker_id)

```

I guess we should not change the PYTHON seed inside the PyTorch dataloarder code.

**Comment by SsnL** [comment]
@dlmacedo Those code shouldn't matter... They are only run on the worker subprocesses. They won't change main process seeds.
